wen-restart/src/wen_restart.rs

```diff
@@ -341,14 +360,100 @@ pub(crate) fn find_heaviest_fork(
     Ok((heaviest_fork_slot, heaviest_fork_bankhash))
 }
 
+// Given the agreed upon slot, make corresponding bank root and generate snapshot.
+fn generate_snapshot(
+    bank_forks: Arc<RwLock<BankForks>>,
+    snapshot_config: &SnapshotConfig,
+    accounts_background_request_sender: &AbsRequestSender,
+    genesis_config_hash: Hash,
+    new_root_slot: Slot,
+    exit: &AtomicBool,
+) -> Result<GenerateSnapshotRecord> {
+    let new_root_bank;
+    {
+        let mut my_bank_forks = bank_forks.write().unwrap();
+        let old_root_bank = my_bank_forks.root_bank();
+        old_root_bank.register_hard_fork(new_root_slot);
+        // new_root_slot is guaranteed to have a bank in bank_forks, it's checked in
+        // find_bankhash_of_heaviest_fork().
+        new_root_bank = my_bank_forks.get(new_root_slot).unwrap();
+        // Only root banks can have snapshots generated.
+        let mut banks = vec![&new_root_bank];
+        let parents = new_root_bank.parents();
+        banks.extend(parents.iter());
+
+        let _ = my_bank_forks.send_eah_request_if_needed(
+            new_root_slot,
+            &banks,
+            accounts_background_request_sender,
+        )?;
+    }
+    // There can't be more than one EAH calculation in progress. If new_root is generated
+    // within the EAH window (1/4 epoch to 3/4 epoch), the following function will wait for
+    // EAH calculation to finish. So if we trigger another EAH when generating snapshots
+    // we won't hit a panic.
+    let _ = new_root_bank.get_epoch_accounts_hash_to_serialize();
+    // Even though generating incremental snapshot is faster, it involves finding a full
+    // snapshot to use as base, so the logic is more complicated. For now we always generate
+    // a full snapshot, even if one already exists, because after adding a hard fork the
+    // bankhash will change.
+    if let Some(slot) =
+        get_highest_full_snapshot_archive_slot(snapshot_config.full_snapshot_archives_dir.clone())
+    {
+        if slot > new_root_slot {
+            warn!(
+                "The new root slot {new_root_slot} should be no smaller than all previous
+                  roots, but the highest full snapshot archive slot is {slot}"
+            );
+        }
```

Comment:
That makes sense, so I guess I can check and abort here if somehow we found a snapshot from future.

However, I don't understand why we can't use the last root. Let's say somehow after root slot 135 all validators just start to crazily fork, so no slot past 135 ever gets enough votes to be the new root slot. Can I just set hard fork on 135, rehash, and generate a new snapshot?

Reply:
Hmm, now that I think about it, it's almost impossible that we get stuck on a rooted slot. Because rooted means it reached max lockout, there must be some nodes behind it which already has almost impossible to expire lockouts.

The only exception is that somehow the algorithm screwed up the "root" conception and many validators suddenly rooted a future slot which shouldn't even be confirmed, or when >2/3 stake is acting abnormally. In both cases we probably shouldn't rely on optimistic restart any more. Instead we should manually inspect and fix the code/situation.

I'll change the requirement to only allow snapshot on a non-rooted slot.

@carllin any objections?

Reply:
I suppose I should also bail out if I can't find any full snapshots at all? This shouldn't happen unless the operator is actively playing with the snapshot directory I guess? Otherwise the validator wouldn't be able to boot at all.

Reply:
> However, I don't understand why we can't use the last root. Let's say somehow after root slot 135 all validators just start to crazily fork, so no slot past 135 ever gets enough votes to be the new root slot. Can I just set hard fork on 135, rehash, and generate a new snapshot

This seems fine

> By requiring a later slot, we can remove a lot of edge cases. Even this reuse-the-same-slot-as-an-existing-snapshot edge case requires some care: removing any existing bank snapshot dir, and handling the old snapshot archive (moving or deleting).

If this wen-restart algorithm decides that indeed the latest snapshot is the latest confirmed slot (no future optimistic slots), seems like we don't even have to regenerate the snapshot, just reuse it

>  is it even possible that we could somehow root slot X and then never optimistically confirm slot X+1 due to forking

Yes this is possible

Reply:
Chatted with Carl a bit, even though the case where we decide on a new_root_slot < last_full_snapshotted_slot is possible, it's probably better to manual inspect in this case. So for now in wen_restart we will not handle this case. It's also possible new_root_slot == last_full_snapshotted_slot, but that is rare enough it's okay not to handle that automatically for now, so we can make the code simpler. We can revisit later if we need it.

Reply:
Changed the logic to always write incremental snapshot, see if this looks better? I'll add more tests tonight or tomorrow.

Reply:
Added more tests, ready for another look.